update winograd convolution WTI computation

CAHEK7 opened the pull request:
Update winograd convolution WTI computation according to the latest measurements and documentation.

The most significant changes are:
- renamed vars to match the documentation
- adjusted some the formulas
- added more precise cycle-related constants

It's an initial PR which updates only current 2x3g1 convolution Wti.
The same code will be reused for groups computation and for the common case of convolution (in separated PRs).

New and old wti computation methods comparison:
[wti-graph.xlsx](https://github.com/ROCmSoftwarePlatform/MIOpen/files/11805547/wti-graph.xlsx)

CAHEK7 committed:
update winograd convolution WTI computation

according to the latest measurements and documentation

Kirpich30000 commented on src/solver/conv_winoRxS.cpp:
```diff
@@ -413,115 +413,213 @@ ConvBinWinoRxS<Winodata, Winofilter>::Search(const ConvolutionContext& ctx,
-    const size_t DATATYPE_BITS;    // S
-    const size_t n_groups;         // BQ ~compute units
-    const bool out_of_model_scope; // Shader model produces unreliable results.
+    static_assert(IS2X3 || IS3X2, "Unsupported filter configuration");
```
IS2X3 == true && IS3X2 == true is also unsupported configuration

CAHEK7 replied:
Sure, only IS2X3 or IS3X2 are supported.
IS2X3 and IS3X2 are mutual exclusive and only IS2X3 can be true or only IS3X2 can be true:
`#define IS2X3 (Winodata == 2 && Winofilter == 3)`
`#define IS3X2 (Winodata == 3 && Winofilter == 2)`
Looks like everything is good here.

CAHEK7 replied:
I'll get rid of this static assert here and move it inside convolution.
Moreover we've decided to remove template parameters, so it can't be static anymore.

CAHEK7 commented:
There are 62 (31 fp1 and 31 fp32) cases where the new algorithm predicted low cycle count and wti returned `-1`.
[wti-disabled-cases.xlsx](https://github.com/ROCmSoftwarePlatform/MIOpen/files/11817034/wti-disabled-cases.xlsx)

Kirpich30000 commented on src/solver/conv_winoRxS.cpp:
```diff
@@ -413,115 +413,213 @@ ConvBinWinoRxS<Winodata, Winofilter>::Search(const ConvolutionContext& ctx,
-        const auto WTI_predicted           = direct_convolution_macs /
-                                   static_cast<double>(N_MACS_PER_CU_PER_CLOCK) /
-                                   static_cast<double>(n_groups) / GUI_predicted; // similar to BW
+        const size_t macs_per_cu_per_clock = 64 * (is_fp16 ? 2 : 1);
```
Only true for gfx10 and gfx906 and earlier. However the fix should consider usage of the predicted wti.

CAHEK7 replied:
I can add more precise calculation based on GPU architecture, but I need some sort of documentation here and we have to decide what would be in a default case.
As far as I know, for gfx11 it should be multiplied by 2.

CAHEK7 replied:
We decided to keep it as it is right now.

Kirpich30000 commented on src/solver/conv_winoRxS.cpp:
```diff
@@ -413,115 +413,213 @@ ConvBinWinoRxS<Winodata, Winofilter>::Search(const ConvolutionContext& ctx,
-                                   static_cast<double>(n_groups) / GUI_predicted; // similar to BW
+        const size_t macs_per_cu_per_clock = 64 * (is_fp16 ? 2 : 1);
+
+        const auto WTI_predicted =
+            direct_convolution_macs /
+            (static_cast<double>(macs_per_cu_per_clock * n_groups) * cycles_predicted);
```
Consider following approach
```
ideal_direct_cycles = direct_convolution_macs / n_CU / macs_per_cu_per_clock;
WTI_predicted = ideal_direct_cycles / cycles_predicted;
```

CAHEK7 replied:
In terms of current algorithm application, changing n_groups to n_CU does not change anything, since right not it is applied to only to the case where `G==1` and `n_groups==n_CU`.
So it doesn't require any additional testing for this PR.

CAHEK7 committed:
fix PR requests

removed template parameter
adjusted final WTI computation
removed most of the class variables

CAHEK7 committed:
simplified code by removing extra class-scoped variables

Kirpich30000 commented on src/solver/conv_winoRxS.cpp:
```diff
@@ -410,118 +410,205 @@ ConvBinWinoRxS<Winodata, Winofilter>::Search(const ConvolutionContext& ctx,
     {
         // Computations do not support negative padding.
         // Negative padding is not applicable, so let use simple assert here.
         assert(pad_h >= 0 && pad_w >= 0);
```
I believe this assert is not needed. Only the padding parity (even or odd) affects granularity.

CAHEK7 replied:
Ok, I've just kept it from the initial version of this class. Can remove it if it seems more correct.

Basically, I don't even know why we initially put this assert here instead of moving this condition it into `out_of_scope` variable and just ignore this algorithm for negative padding and decided to crash with assertion.

Kirpich30000 replied:
We could remove it

CAHEK7 committed:
remove outdated assertion

Kirpich30000 reviewed (approved):
LGTM

junliume commented on src/solver/conv_winoRxS.cpp:
```diff
@@ -410,118 +410,201 @@ ConvBinWinoRxS<Winodata, Winofilter>::Search(const ConvolutionContext& ctx,
+        {
+            n_groups       = n_CU;
+            n_works_per_CU = Ceil(n_works, n_CU);
+            return;
```
@CAHEK7 hip tidy issue on this line:
```
[2023-06-24T07:12:45.296Z] /var/jenkins/workspace/MLLibs_MIOpen_develop/src/solver/conv_winoRxS.cpp:521:13: error: 1 uninitialized field at the end of the constructor call [clang-analyzer-optin.cplusplus.UninitializedObject,-warnings-as-errors]
[2023-06-24T07:12:45.296Z]             return;
[2023-06-24T07:12:45.296Z]             ^

CAHEK7 replied:
Addressed in #2224

junliume commented on src/solver/conv_winoRxS.cpp:
```diff
@@ -410,118 +410,201 @@ ConvBinWinoRxS<Winodata, Winofilter>::Search(const ConvolutionContext& ctx,
-                                   static_cast<double>(n_groups) / GUI_predicted; // similar to BW
+        /// \todo Adjust for different architectures, probably it should be an external source for
+        /// this metric
+        const size_t macs_per_cu_per_clock = 64 * (is_fp16 ? 2 : 1);
```
@CAHEK7 hip tidy issue with this line:
```
[2023-06-24T07:12:45.297Z] /var/jenkins/workspace/MLLibs_MIOpen_develop/src/solver/conv_winoRxS.cpp:602:46: error: performing an implicit widening conversion to type 'const size_t' (aka 'const unsigned long') of a multiplication performed in type 'int' [bugprone-implicit-widening-of-multiplication-result,-warnings-as-errors]

[2023-06-24T07:12:45.297Z]         const size_t macs_per_cu_per_clock = 64 * (is_fp16 ? 2 : 1);

[2023-06-24T07:12:45.297Z]                                              ^

CAHEK7 replied:
Addressed in #2224

junliume commented:
@CAHEK7 my mistake thinking this PR passed CI (the green tick). Unfortunately I have to revert this PR, due to reasons claimed above. Could you please re-work this PR to fix the issue? For new PR I will help to get CI through. Thanks!
